### ISSUE TITLE
Tests for dataset context manager subscriptions are flaky

### ISSUE THREAD

**Issue by WilliamHPNielsen** [QCoDeS/Qcodes#1124]
@QCoDeS/core know what I'm talking about. It's annoying as hell and might point to some scary underlying problem. We should really try to figure out what is going on.

**Comment by sohailc** [comment]
http://hypothesis.readthedocs.io/en/latest/reproducing.html

**Comment by sohailc** [comment]
As stated in PR 1131, we "fixed" the test_subscriptions test by explicitly calling flush_to_database explicitly instead of waiting a period of time before calling "add_result". Even if this fixes the "test_subscriptions" test, there is still some underlying problem because of which data is not flushed to the database at expected times. This still needs to be investigated.

**Comment by jenshnielsen** [comment]
In this travis run https://travis-ci.org/QCoDeS/Qcodes/builds/390728436?utm_source=github_status&utm_medium=notification test_subscriptions_getting_all_points failed again even with this fix

**Comment by WilliamHPNielsen** [comment]
And here `test_subscriptions` failed: https://travis-ci.org/QCoDeS/Qcodes/builds/391788144?utm_source=github_status&utm_medium=notification

-Darn!

**Comment by astafan8** [comment]
Thank you for collecting the failed runs. I will look into that soon...

**Comment by astafan8** [comment]
as @jenshnielsen suggested, perhaps, the problem is in the resolution of `monotonic()` calls. It might help to substitute them with [performance counter](https://docs.python.org/3/library/time.html#time.perf_counter).

**Comment by WilliamHPNielsen** [comment]
@astafan8 I think it's worth trying. Do you want to give it a go?

**Comment by astafan8** [comment]
sure, i'm on it :) I am still the one whom this issue is assigned.

**Comment by astafan8** [comment]
@jenshnielsen @WilliamHPNielsen I think i've found the problem. [I've also implemented a solution that works](https://github.com/astafan8/Qcodes/commit/a68f7256de232adbccae24da1e86402cd375c6d6). But I think there is some problem with responsibilities between objects. Also, I got a general impression that subscribers are implemented in not a clean way, and may be potentially thread unsafe. I'll expand on it later in a PR or even in a new issue.
